workspaces/controller/internal/controller/workspace_controller.go

```diff
@@ -913,6 +996,105 @@ func generateService(workspace *kubefloworgv1beta1.Workspace, imageConfigSpec ku
 	return service, nil
 }
 
+// buildHTTPRoute creates an HTTPRoute for a given port configuration
+func buildHTTPRoute(
+	serviceHost string,
+	matchUriPrefix string,
+	port kubefloworgv1beta1.ImagePort,
+	podTemplatePort kubefloworgv1beta1.WorkspaceKindPort,
+) *networkingv1.HTTPRoute {
+
+	httpRoute := &networkingv1.HTTPRoute{
+		Headers: &networkingv1.Headers{
+			Request: &networkingv1.Headers_HeaderOperations{},
+		},
+		Rewrite: &networkingv1.HTTPRewrite{},
+		Match: []*networkingv1.HTTPMatchRequest{
+			{
+				Uri: &networkingv1.StringMatch{
+					MatchType: &networkingv1.StringMatch_Prefix{
+						Prefix: matchUriPrefix,
+					},
+				},
+			},
+		},
+		Route: []*networkingv1.HTTPRouteDestination{
+			{
+				Destination: &networkingv1.Destination{
+					Host: serviceHost,
+					Port: &networkingv1.PortSelector{
+						Number: uint32(port.Port), //nolint:gosec
+					},
+				},
+			},
+		},
+	}
+
+	if !ptr.Deref(podTemplatePort.HTTPProxy.RemovePathPrefix, false) {
+		httpRoute.Rewrite.Uri = matchUriPrefix
+	}
+
+	if podTemplatePort.HTTPProxy.RequestHeaders != nil {
+		httpRoute.Headers.Request.Set = podTemplatePort.HTTPProxy.RequestHeaders.Set
+		httpRoute.Headers.Request.Add = podTemplatePort.HTTPProxy.RequestHeaders.Add
+		httpRoute.Headers.Request.Remove = podTemplatePort.HTTPProxy.RequestHeaders.Remove
+	}
+
+	return httpRoute
+}
+
+// generateVirtualService generates a VirtualService for a Workspace
+func generateVirtualService(workspace *kubefloworgv1beta1.Workspace, workspaceKind *kubefloworgv1beta1.WorkspaceKind, serviceName string, imageConfigSpec kubefloworgv1beta1.ImageConfigSpec, cfg *config.EnvConfig) (*istiov1.VirtualService, error) {
+	if cfg.IstioGateway == "" {
+		return nil, fmt.Errorf("ISTIO_GATEWAY environment variable is not set")
+	}
+
+	// NOTE: the name prefix is used to generate a unique name for the VirtualService
+	namePrefix := generateNamePrefix(workspace.Name, maxVirtualServiceNameLength)
+
+	currentPodTemplatePortsMap := make(map[kubefloworgv1beta1.PortId]kubefloworgv1beta1.WorkspaceKindPort)
+	for _, port := range workspaceKind.Spec.PodTemplate.Ports {
+		currentPodTemplatePortsMap[port.Id] = port
+	}
+
+	serviceHost := fmt.Sprintf("%s.%s.svc.%s", serviceName, workspace.Namespace, cfg.ClusterDomain)
+
+	httpRoutes := []*networkingv1.HTTPRoute{}
+	for _, port := range imageConfigSpec.Ports {
+		// Guard: skip if port doesn't exist in current pod template
```

Comment:
To make it clear why we don't error, we should say:

```suggestion
		// silently ignore port ids not defined in the workspace kind
		// NOTE: this should not be possible as the webhook blocks undefined ports
```